classes/ActionScheduler_wpCommentCleaner.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * Class ActionScheduler_WPCommentCleaner
+ *
+ * @since 3.0.0
+ */
+class ActionScheduler_WPCommentCleaner {
+
+	/**
+	 * Post migration hook used to cleanup the WP comment table.
+	 *
+	 * @var string
+	 */
+	protected static $cleanup_hook = 'action_scheduler/cleanup_wp_comment_logs';
+
+	/**
+	 * An instance of the ActionScheduler_wpCommentLogger class to interact with the comments table.
+	 *
+	 * This instance should only be used as an interface. It should not be initialized.
+	 *
+	 * @var ActionScheduler_wpCommentLogger
+	 */
+	protected static $wp_comment_logger = null;
+
+	/**
+	 * Initialize the class and attach callbacks.
+	 */
+	public static function init() {
+		self::$wp_comment_logger = new ActionScheduler_wpCommentLogger();
+
+		add_action( self::$cleanup_hook, array( __CLASS__, 'delete_all_action_comments' ) );
+
+		// While there are orphaned logs left in the comments table, we need to attach the callbacks which filter comment counts.
+		add_action( 'pre_get_comments', array( self::$wp_comment_logger, 'filter_comment_queries' ), 10, 1 );
+		add_action( 'wp_count_comments', array( self::$wp_comment_logger, 'filter_comment_count' ), 20, 2 ); // run after WC_Comments::wp_count_comments() to make sure we exclude order notes and action logs
+		add_action( 'comment_feed_where', array( self::$wp_comment_logger, 'filter_comment_feed' ), 10, 2 );
+
+		// Action Scheduler may be displayed as a Tools screen or WooCommerce > Status administration screen
+		add_action( 'load-tools_page_action-scheduler', array( __CLASS__, 'print_admin_notices' ) );
+		add_action( 'load-woocommerce_page_wc-status', array( __CLASS__, 'print_admin_notices' ) );
+	}
+
+	/**
+	 * Determines if there are log entries in the wp comments table.
+	 *
+	 * @return boolean Whether there are scheduled action comments in the comments table.
+	 */
+	public static function has_logs() {
+		return (bool) get_comments( array( 'type' => ActionScheduler_wpCommentLogger::TYPE, 'number' => 1, 'fields' => 'ids' ) );
+	}
+
+	/**
+	 * Schedules the WP Post comment table cleanup to run in 6 months if it's not already scheduled.
+	 * Attached to the migration complete hook 'action_scheduler/migration_complete'.
+	 */
+	public static function maybe_schedule_cleanup() {
+		if ( ! as_next_scheduled_action( self::$cleanup_hook ) && self::has_logs() ) {
+			as_schedule_single_action( gmdate( U ) + ( 6 * MONTH_IN_SECONDS ), self::$cleanup_hook );
```

Comment:
6 months seems a bit long. In normal operation AS purges actions after 30 days.

Reply:
The 30 day purge is for completed actions. For failed actions, there is no auto purge. I'd consider this data more akin to failed actions, because we know something is wrong, so I think 6 months is OK. For most sites, it shouldn't be causing issues, and it's relatively easy to address if it is causing issues (i.e. install a plugin to delete orphaned comments, or run a SQL query).